### ISSUE TITLE
What does HistFactory's enableOffsetting do and does it give speed-ups?

### ISSUE THREAD

**Issue by kratsg** [diana-hep/pyhf#254]
# Question

Similar to #253, code like the following

```c++
  RooAbsReal *nll = pdf->createNLL(*data, NumCPU(_nCPU),
				   Constrain(*mc->GetNuisanceParameters()),
				   GlobalObservables(*mc->GetGlobalObservables()),
				   ConditionalObservables(*mc->GetConditionalObservables()),
				   ExternalConstraints(*_externalConstraint));
  if (_fixStarCache) utils::fixRooStarCache( w ); // needs to be done twice.
  nll->enableOffsetting( _nllOffset );
```
is reportedly giving speed-ups, at least for the Higgs folks doing fits there. Any ideas what it does and why?

**Comment by gadatsch** [comment]
The idea is subtracting large NLL values to improve the numeric stability of the fit.

**Comment by kratsg** [comment]
> The idea is subtracting large NLL values to improve the numeric stability of the fit.

Removing or subtracting?

**Comment by matthewfeickert** [comment]
I don't think this is actually that helpful, but I would assume that it is similar to [`RooMinimizer::setOffsetting`](https://root.cern.ch/doc/v612/classRooMinimizer.html#a7a5592e20f80f783e59f557d6d51ac45) which
> Enable[s] internal likelihood offsetting for enhanced numeric precision.

which I think is just doing subtracting. The not clear on the mechanics of how this is done though and this isn't well documented from what I've found in the past.

**Comment by gadatsch** [comment]
Yes, it's just a subtraction, see [here](https://root.cern.ch/doc/master/RooNLLVar_8cxx_source.html#l00404). The idea is that with large NLL values stored, the the precision on the values is decreased and Minuit's VariableMetric steps can be inaccurate resulting in a non-smooth decrease of the edm and a not properly converging fit.
